credentials/src/main/kotlin/web5/sdk/credentials/PresentationExchange.kt

```diff
@@ -20,40 +28,77 @@ public object PresentationExchange {
         "Presentation Definition's Submission Requirements feature is not implemented"
       )
     }
 
-    return presentationDefinition.inputDescriptors
+    presentationDefinition.inputDescriptors
       .filter { !it.constraints.fields.isNullOrEmpty() }
-      .all { inputDescriptorWithFields ->
-        val requiredFields = inputDescriptorWithFields.constraints.fields!!.filter { it.optional != true }
-
-        var satisfied = true
-        for (field in requiredFields) {
-          // we ignore field filters
-          if (field.filterSchema != null) {
-            throw UnsupportedOperationException("Field Filter is not implemented")
-          }
-
-          if (field.path.any { path -> credential.getFieldByJsonPath(path) == null }) {
-            satisfied = false
-            break
-          }
-        }
-        return satisfied
+      .forEach { inputDescriptorWithFields ->
+        validateInputDescriptorsWithFields(inputDescriptorWithFields, vc.payload)
       }
   }
+
+  /**
+   * Validates whether the input descriptors with fields in a Presentation Definition are satisfied by
+   * the payload of a verifiable credential (VC).
+   *
+   * @param inputDescriptorWithFields The input descriptor with fields to be validated.
+   * @param vcPayload The payload of the VC.
+   */
+  private fun validateInputDescriptorsWithFields(
+    inputDescriptorWithFields: InputDescriptorV2,
+    vcPayload: Payload
+  ) {
+    val requiredFields = inputDescriptorWithFields.constraints.fields!!.filter { it.optional != true }
+
+    requiredFields.forEach { field ->
+      val vcPayloadJson = JsonPath.parse(vcPayload.toString())
+        ?: throw PresentationExchangeError("Failed to parse VC $vcPayload as JsonNode")
+
+      val matchedFields = field.path.mapNotNull { path -> vcPayloadJson.read<JsonNode>(path) }
+      if (matchedFields.isEmpty()) {
+        throw PresentationExchangeError("Could not find matching field for path: ${field.path.joinToString()}")
+      }
+
+      when {
+        field.filterSchema != null -> matchedFields.any {vcSatisfiesFieldFilterSchema(it, field.filterSchema!!)}
+        else -> return
+      }
+    }
+  }
+
+  /**
+   * Validates whether a verifiable credential (VC) field value satisfies a JSON schema.
+   *
+   * @param fieldValue The field value of the VC as a JsonNode.
+   * @param schema The JSON schema to validate against.
+   */
+  private fun vcSatisfiesFieldFilterSchema(fieldValue: JsonNode, schema: JsonSchema): Boolean {
+    val validationMessages = schema.validate(fieldValue)
+    when {
+      validationMessages.isEmpty() -> return true
+      else -> throw PresentationExchangeError("validating $fieldValue failed: ${validationMessages.joinToString()}")
+    }
+  }
 }
+
+/**
+ * Custom error class for exceptions related to the Presentation Exchange.
+ *
+ * @param message The error message.
+ */
+public class PresentationExchangeError(message: String) : Error(message)
```

Comment:
Couldn't find where error is defined, is it an exception?

Reply:
`Error` is `java.lang.Error`, I'm actually not sure when to use error vs exception

Reply:
From https://docs.oracle.com/javase%2F7%2Fdocs%2Fapi%2F%2F/java/lang/Error.html

> An Error is a subclass of Throwable that indicates serious problems that a reasonable application should not try to catch.

That sounds really scary 😱 

In other PRs, we've tried to stick subclassing `RuntimeException`, like what's below https://github.com/TBD54566975/web5-kt/blob/02ea4b31a28cf68a4a3f9a9abc745e3273ceee79/dids/src/main/kotlin/web5/sdk/dids/DidIonManager.kt#L720